### ISSUE TITLE
Configurable number of retries in a write-behind store

### ISSUE THREAD

**Issue by GhostInAMachine** [hazelcast/hazelcast#9265]
Right now, writing an item to a write-behind store will be retried indefinitely if there's any exception thrown. The only way to avoid this is to suppress the exception, which is not that elegant. 

Instead, a retry limit option could be introduced to the MapStore configuration.

**Comment by ahmetmircik** [comment]
a similar issue: https://github.com/hazelcast/hazelcast/issues/3197

**Comment by zjedi** [comment]
Maybe the whole retry algorithm should be a pluggable component (code, not just a few parameters). We have intese spikes, which may lead to misleading stats being calculated: the database is overloaded just temporarily, but dirty entries seem to hang around for quite some time after the spike and are cleared in bigger batches.

**Comment by pertsodian** [comment]
I see we have `com.hazelcast.client.impl.connection.nio.WaitStrategy#sleep` which does the calculation for sleep time before attempting a cluster reconnection. It should be possible to extract the logic there for reuse, in this case, to retry the write-behind operation.

Question: do we want to give up ever if the exception keeps happening? Giving up after certain amount of time/attempts = data inconsistency between cache and store?

**Comment by pertsodian** [comment]
@ahmetmircik are you working on this issue? Maybe I can give it a try if you agree with my above comment?

**Comment by ahmetmircik** [comment]
your idea sounds good to me, unassigned myself, issue is yours.

**Comment by pertsodian** [comment]
@ahmetmircik, and maybe @mmedenjak too. I need your input before going too far.

I have looked at `StoreWorker` and `DefaultWriteBehindProcessor`.

If I understand it correctly, the current retry logic is:
* In each run, `StoreWorker` collects to-be-stored entries from the queues for each partition (according to lastStoreTime+writeDelay vs now), and attempts to store the entries belong to the partitions that the member owns
* Depending on `writeBatchSize` / `writeCoalescing` / actual operation type, `callSingleStoreWithListeners` or `callBatchStoreWithListeners` will be called
* If `callBatchStoreWithListeners` fails, `callSingleStoreWithListeners` will be called for each entry in the batch
* These two types of call will be attempted 3 times, 1 second apart before it returns. These are not configurable as of now, but maybe they don't need to be.
* All the failed entries will be collected and returned back to `StoreWorker`
* `StoreWorker` add the failed entries back to the front of the queue, and end the current run.
* Next run is scheduled after 1 second delay.

If we sleep before putting the failed entries back to the queue, the whole write-behind flow for current map will be impacted, since there is only one `StoreWorker` per map.

So I am thinking to add more attributes to `DelayedEntry` to tell when it should be attempted to store, e.g. lastFailureTS and waitTime
* New entries will have the attributes set accordingly by default so that there is no waiting, if it is picked up by `StoreWorker`
* Failed entries will have lastFailureTS = now, and waitTime calculated from previous waitTime and other waiting config specified by `MapStoreConfig` (this calculation logic might reuse from `WaitStrategy`)
* `DefaultWriteBehindProcessor` to check these attributes before attempt to store them. If the time is not right, then don't attempt and just return as failed immediately.

The above will help to not block the write-behind of the entire map, if some keys fail.
Does that sound ok to implement?

If yes, then I need answer to one more question.
Let's say <K1, V1> fails, and put back to the front of the queue. After a few more failed attempts, the wait time has accumulated to minutes, for example. Then <K1, V2> is now also in the queue, and also picked up by `StoreWorker` (now time is higher than last store time + write delay).
If `writeCoalescing` is enabled, <K1, V1> will be dropped, and we persist <K1, V2> only. Should <K1, V2> inherit the wait time from <K1, V1>, or should it reset?
Implementation wise, both should be doable I think.

**Comment by Holmistr** [comment]
Thanks @pertsodian for looking into this. Just as a reminder, we should try addressing requirements from https://github.com/hazelcast/hazelcast/issues/3197#issue-39732662 somehow as well.

**Comment by pertsodian** [comment]
@Holmistr Yeah the waiting time calculation will be configuration, and will be similar to how we configure cluster reconnection (See [ConnectionRetryConfig](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/config/ConnectionRetryConfig.java))
* initial backoff
* max backoff
* multiplier
* jitter

User can set multiplier > 1 if they want exponential backoff.

**Comment by Holmistr** [comment]
@pertsodian Sorry for late responses and no feedback so far. I'm pushing again :)

**Comment by ahmetmircik** [comment]
@pertsodian sorry for late response, i haven't been around lately.

Your design for the solution made sense to me.

Regarding this:
> Should <K1, V2> inherit the wait time from <K1, V1>, or should it reset?

Not sure about this one but we can inherit the time if it failed recently.

**Comment by pertsodian** [comment]
Hello all, I am so sorry I could not make time to work on this task with proper attention. I will unassign myself so that others can pick. Thank you for your understanding.
